src/services/deep-research/DeepResearchService.ts

```diff
@@ -0,0 +1,930 @@
+	researchQuerySchema,
+	researchTaskSchema,
+} from "./types"
+import { RecursiveCharacterTextSplitter, encoder } from "./TextSplitter"
+
+type DeepResearchServiceStatus = "idle" | "followUp" | "research" | "done" | "aborted"
+
+export class DeepResearchService {
+	public readonly taskId: string
+	public readonly providerId: string
+	public readonly providerApiKey: string
+	public readonly firecrawlApiKey: string
+	public readonly modelId: string
+	public readonly breadth: number
+	public readonly depth: number
+	public readonly concurrency: number
+
+	private providerRef: WeakRef<ClineProvider>
+	private firecrawl: FirecrawlApp
+	private openai: OpenAI
+	private _status: DeepResearchServiceStatus = "idle"
+
+	private inquiry: ResearchInquiry = { followUps: [], responses: [] }
+	private progress: ResearchProgress = { expectedQueries: 0, completedQueries: 0, progressPercentage: 0 }
+
+	private tokenUsage: ResearchTokenUsage = { inTokens: 0, outTokens: 0, totalTokens: 0 }
+	private output: ResearchOutput[] = []
+	private messages: ChatCompletionMessageParam[] = []
+
+	constructor(
+		{
+			providerId,
+			providerApiKey,
+			firecrawlApiKey,
+			modelId,
+			breadth,
+			depth,
+			concurrency,
+		}: ResearchTaskPayload["session"],
+		clineProvider: ClineProvider,
+	) {
+		this.taskId = crypto.randomUUID()
+		this.providerId = providerId
+		this.providerApiKey = providerApiKey
+		this.firecrawlApiKey = firecrawlApiKey
+		this.modelId = modelId
+		this.breadth = breadth
+		this.depth = depth
+		this.concurrency = concurrency
+
+		this.providerRef = new WeakRef(clineProvider)
+
+		this.firecrawl = new FirecrawlApp({ apiKey: firecrawlApiKey })
+
+		this.openai = new OpenAI({
+			baseURL: providerId === "openrouter" ? "https://openrouter.ai/api/v1" : undefined,
+			apiKey: providerApiKey,
+		})
+	}
+
+	/**
+	 * Prompts
+	 */
+
+	private researchSystemPrompt() {
+		const now = new Date().toISOString()
+
+		return this.trimPrompt(`
+            You are an expert researcher. Today is ${now}. Follow these instructions when responding:
+            - You may be asked to research subjects that is after your knowledge cutoff, assume the user is right when presented with news.
+            - The user is a highly experienced analyst, no need to simplify it, be as detailed as possible and make sure your response is correct.
+            - Be highly organized.
+            - Suggest solutions that I didn't think about.
+            - Be proactive and anticipate my needs.
+            - Treat me as an expert in all subject matter.
+            - Mistakes erode my trust, so be accurate and thorough.
+            - Provide detailed explanations, I'm comfortable with lots of detail.
+            - Value good arguments over authorities, the source is irrelevant.
+            - Consider new technologies and contrarian ideas, not just the conventional wisdom.
+            - You may use high levels of speculation or prediction, just flag it for me.
```

Comment:
Might be worth making this consistent between "me"/"I" and "the user"

Reply:
Ah, that is pretty weird. This was unmodified from the original prompt, but I just made it more consistent.